### ISSUE TITLE
Mocking and injecting Reflux store with Rewire

### ISSUE THREAD

**Issue by SiCurious** [spoike/refluxjs#300]
I'm attempting to mock out and inject a Reflux store into a Component for the purpose of testing:

**store.js**
```javascript
var Reflux = require('reflux');
var key = 0;

module.exports = Reflux.createStore({
  getInitialState: function() {
    this.list = [{
      key: key++,
      label: 'Initial state'
    }];
    return this.list;
  }
});
```
**component.js**
```javascript
var React = require('react');
var Reflux = require('reflux');
var Store = require('./../store');

module.exports = React.createClass({

  mixins: [Reflux.connect(Store, 'list')],

  render: function() {

    var listItems = this.state.list.map(function(listItem) {
        return ( <li>{listItem.label}</li> );
    });
    return ( <div><ul>{listItems}</ul></div> );
  }
});
```
**component-spec.js**
```javascript
var React = require('react/addons');
var Reflux = require('reflux');
var TestUtils = React.addons.TestUtils;
var rewire = require('rewire');
var Component = rewire('../../../app/component');
var renderedComponent;

describe('Component', function() {

  var StoreMock = Reflux.createStore({
    getInitialState: function() {
      this.list = [{
        key: 'key_1',
        label: 'Mock initial state'
      }];
      return this.list;
    }
  });

  Component.__set__('Store', StoreMock);

  beforeEach(function() {
    renderedComponent = TestUtils.renderIntoDocument(
      <Component />
    );
  });

  it('should have an initial state', function() {
    var liElements = TestUtils.scryRenderedDOMComponentsWithTag(renderedComponent, 'li');
    expect(liElements[0].getDOMNode().innerHTML).toBe('Mock initial state');
  });
});
```

In this instance, my test fails, expecting 'Initial state' to be 'Mock initial state'. 

Am I doing something wrong in attempting to mock out the store in this way? It seems that the component is still connecting to the real store, even though I have rewired it in the test spec.

Please let me know if you need any more information.

**Comment by p-equis** [comment]
I ran into this - but it's not a problem with reflux. More like a weakness with the model that rewire uses to inject dependencies.

By the time you make your `Component.__set__` call, the `Store` has *already been used*. It's mixed in by React, and `Store` is used before the component is even exported. Since the rewire call happens after the export, it cannot affect what was mixed in to the component.

Because of this problem I switched to using [proxyquire](https://github.com/thlorenz/proxyquire), which actually allows you to overwrite what is returned from `require` calls.
